packages/microcosm/src/addons/model.js

```diff
@@ -14,116 +15,186 @@ function isCallable(binding: *): boolean {
   return binding && typeof binding.call === 'function'
 }
 
-function invoke(binding: *, repo: Microcosm, scope: Object): * {
-  if (isCallable(binding)) {
-    return binding.call(scope, repo.state, repo)
-  }
-
-  return binding
-}
+type Answer = { [string]: * }
 
 export default class Model extends Emitter {
-  repo: Microcosm
-  scope: Object
-  bindings: { [key: string]: * }
-  subscriptions: { [key: string]: * }
-  revision: number
-  value: *
+  _scope: Object
+  _repo: Microcosm
+  _bindings: *
+  _subscriptions: { [string]: * }
+  _pendingUpdate: ?Job
+  _publish: () => void
+  _patch: Answer
+  _watching: Boolean
+
+  value: Answer
 
   constructor(repo: Microcosm, scope: Object) {
     super()
 
-    this.repo = repo
-    this.scope = scope
-    this.bindings = {}
-    this.subscriptions = {}
+    this._repo = repo
+    this._scope = scope
+    this._bindings = {}
+    this._subscriptions = {}
+    this._patch = {}
+    this._publish = this._publish.bind(this)
+    this._pendingUpdate = null
+    this._watching = false
+
     this.value = {}
-    this.revision = 0
+  }
+
+  bind(bindings: Object) {
+    let nextBindings = null
+    let recompute = false
 
-    this.repo.on('change', this.compute, this)
+    for (var key in bindings) {
+      let callback = bindings[key]
+
+      if (isObservable(callback)) {
+        // Presenters support observables. Updates to observables
+        // assign to a single key/value pair
+        this._track(key, callback)
+      } else if (isCallable(callback)) {
+        // Presenters will "call" anything that looks like it's a function.
+        // It does this by checking the "call" property.
+        //
+        // If the binding is new, rebind! Otherwise we don't want to
+        // recalculate this model if nothing has changed.
+        if (this._bindings[key] !== callback) {
+          recompute = true
+          nextBindings = nextBindings || {}
+          nextBindings[key] = callback
+        }
+      } else if (this.value[key] !== callback) {
+        // Otherwise we just have a primitive value. If the value is
+        // different, push the key/value pair into the next patch and
+        // queue a recompute.
+        recompute = true
+        this._patch[key] = callback
+      }
+    }
```

Comment:
Man, this is too many use cases! I'd really like to drop the callable interface at some point and just use primitive values or observables.

Reply:
There's a pretty tremendous perf boost if we drop callables too, because it means we can stop listening to the repo naively and "push" new values in as needed - ideally batched.